File: wire-compiler/src/main/java/com/squareup/wire/model/Pruner.java
/*
 * Copyright (C) 2015 Square, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.squareup.wire.model;

import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Deque;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

/** Removes unused types and services. */
public final class Pruner {
  /** Homogeneous identifiers including type names, service names, and RPC names. */
  final Set<String> marks = new LinkedHashSet<String>();

  /** Identifiers whose immediate dependencies have not yet been marked. */
  final Deque<String> queue = new ArrayDeque<String>();

  /**
   * Returns a new root set that contains only the types in {@code roots} and their transitive
   * dependencies.
   *
   * @param roots a set of identifiers to retain, which may be fully qualified type names, fully
   *     qualified service names, or service RPCs like {@code package.ServiceName#MethodName}.
   */
  public List<WireProtoFile> retainRoots(List<WireProtoFile> protoFiles, Set<String> roots) {
    if (roots.isEmpty()) throw new IllegalArgumentException();
    if (!marks.isEmpty()) throw new IllegalStateException();

    Map<String, WireType> typesIndex = buildTypesIndex(protoFiles);
    Map<String, WireService> servicesIndex = buildServicesIndex(protoFiles);

    // Mark and enqueue the roots.
    for (String s : roots) {
      mark(s);
    }

    // Extensions and options are also roots.
    for (WireProtoFile protoFile : protoFiles) {
      for (WireExtend extend : protoFile.wireExtends()) {
        markExtend(extend);
      }
      markOptions(protoFile.options());
    }

    // Mark everything reachable by what's enqueued, queueing new things as we go.
    for (String name; (name = queue.poll()) != null;) {
      if (ProtoTypeName.getScalar(name) != null) {
        continue; // Skip scalar types.
      }

      WireType type = typesIndex.get(name);
      if (type != null) {
        markType(type);
        continue;
      }

      WireService service = servicesIndex.get(name);
      if (service != null) {
        markService(service);
        continue;
      }

      // If the root set contains a method name like 'Service#Method', only that RPC is marked.
      int hash = name.indexOf('#');
      if (hash != -1) {
        String serviceName = name.substring(0, hash);
        String rpcName = name.substring(hash + 1);
        WireService partialService = servicesIndex.get(serviceName);
        if (partialService != null) {
          WireRpc rpc = partialService.rpc(rpcName);
          if (rpc != null) {
            markOptions(partialService.options());
            markRpc(rpc);
            continue;
          }
        }
      }

      throw new IllegalArgumentException("Unexpected type: " + name);
    }

    List<WireProtoFile> retained = new ArrayList<WireProtoFile>();
    for (WireProtoFile protoFile : protoFiles) {
      retained.add(protoFile.retainAll(marks));
    }

    return ImmutableList.copyOf(retained);
  }

  private static Map<String, WireType> buildTypesIndex(Collection<WireProtoFile> protoFiles) {
    Map<String, WireType> result = new LinkedHashMap<String, WireType>();
    for (WireProtoFile protoFile : protoFiles) {
      for (WireType type : protoFile.types()) {
        index(result, type);
      }
    }
    return ImmutableMap.copyOf(result);
  }

  private static void index(Map<String, WireType> typesByName, WireType type) {
    typesByName.put(type.protoTypeName().toString(), type);
    for (WireType nested : type.nestedTypes()) {
      index(typesByName, nested);
    }
  }

  private static Map<String, WireService> buildServicesIndex(Collection<WireProtoFile> protoFiles) {
    Map<String, WireService> result = new LinkedHashMap<String, WireService>();
    for (WireProtoFile protoFile : protoFiles) {
      for (WireService service : protoFile.services()) {
        result.put(service.protoTypeName().toString(), service);
      }
    }
    return ImmutableMap.copyOf(result);
  }

  private void mark(ProtoTypeName typeName) {
    mark(typeName.toString());
  }

  private void mark(String identifier) {
    if (marks.add(identifier)) {
      queue.add(identifier); // The transitive dependencies of this identifier must be visited.
    }
  }

  private void markExtend(WireExtend extend) {
    mark(extend.protoTypeName());
    markFields(extend.fields());
  }

  private void markType(WireType type) {
    markOptions(type.options());

    ProtoTypeName enclosingTypeName = type.protoTypeName().enclosingTypeName();
    if (enclosingTypeName != null) {
      mark(enclosingTypeName);
    }

    for (WireType nestedType : type.nestedTypes()) {
      mark(nestedType.protoTypeName());
    }

    if (type instanceof WireMessage) {
      markMessage((WireMessage) type);
    } else if (type instanceof WireEnum) {
      markEnum((WireEnum) type);
    }
  }

  private void markMessage(WireMessage message) {
    markFields(message.fields());
    for (WireOneOf oneOf : message.oneOfs()) {
      markFields(oneOf.fields());
    }
  }

  private void markEnum(WireEnum wireEnum) {
    markOptions(wireEnum.options());
    for (WireEnumConstant constant : wireEnum.constants()) {
      markOptions(constant.options());
    }
  }

  private void markFields(List<WireField> fields) {
    for (WireField field : fields) {
      markField(field);
    }
  }

  private void markField(WireField field) {
    markOptions(field.options());
    mark(field.type());
  }

  private void markOptions(Options options) {
    for (WireField field : options.fields()) {
      markField(field);
    }
  }

  private void markService(WireService service) {
    markOptions(service.options());
    for (WireRpc rpc : service.rpcs()) {
      markRpc(rpc);
    }
  }

  private void markRpc(WireRpc rpc) {
    markOptions(rpc.options());
    mark(rpc.requestType());
    mark(rpc.responseType());
  }
}

File: wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/Pruner.kt
/*
 * Copyright (C) 2015 Square, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.squareup.wire.schema

import com.squareup.wire.schema.ProtoMember.Companion.get
import com.squareup.wire.schema.internal.mutableQueueOf

/**
 * Creates a new schema that contains only the types selected by the pruning rules, including their
 * transitive dependencies.
 */
class Pruner(
  private val schema: Schema,
  private val pruningRules: PruningRules,
) {
  private val marks = MarkSet(pruningRules)

  /**
   * [types][ProtoType] and [members][ProtoMember] whose immediate dependencies have not
   * yet been visited.
   */
  private val queue = mutableQueueOf<Any>()

  fun prune(): Schema {
    markRoots()
    markReachable()

    val retained = retainImports(retainAll(schema, marks))

    return Schema(retained)
  }

  private fun retainAll(
    schema: Schema,
    marks: MarkSet,
  ): List<ProtoFile> {
    return schema.protoFiles.map { protoFile ->
      protoFile.retainAll(schema, marks)
    }
  }

  private fun retainImports(protoFiles: List<ProtoFile>): List<ProtoFile> {
    val schema = Schema(protoFiles)
    return protoFiles.map { protoFile ->
      protoFile.retainImports(schema)
    }
  }

  private fun markRoots() {
    for (protoFile in schema.protoFiles) {
      markRoots(protoFile)
    }
  }

  private fun markRoots(protoFile: ProtoFile) {
    for (type in protoFile.types) {
      markRootsIncludingNested(type)
    }
    for (service in protoFile.services) {
      markRoots(service.type)
    }
  }

  private fun markRootsIncludingNested(type: Type) {
    markRoots(type.type)

    for (nested in type.nestedTypes) {
      markRootsIncludingNested(nested)
    }
  }

  private fun markRoots(protoType: ProtoType) {
    if (pruningRules.isRoot(protoType)) {
      marks.root(protoType)
      queue.add(protoType)
      return
    }

    // The top-level type isn't a root, search for root members inside.
    for (reachable in reachableObjects(protoType)) {
      if (reachable !is ProtoMember) continue
      if (!isRetainedVersion(reachable)) continue
      if (pruningRules.isRoot(reachable)) {
        marks.root(reachable)
        marks.mark(reachable.type) // Consider this type as visited.
        queue.add(reachable)
      }
    }
  }

  /** Returns true if this member survives `since` and `until` pruning. */
  private fun isRetainedVersion(protoMember: ProtoMember): Boolean {
    val member = protoMember.member
    return when (val type = schema.getType(protoMember.type)) {
      is MessageType -> {
        val field = type.field(member) ?: type.extensionField(member)
        if (field != null) {
          pruningRules.isFieldRetainedVersion(field.options)
        } else {
          pruningRules.isFieldRetainedVersion(type.oneOf(member)!!.options)
        }
      }
      is EnumType -> {
        val enumConstant = type.constant(member)!!
        pruningRules.isEnumConstantRetainedVersion(enumConstant.options)
      }
      else -> true
    }
  }

  /**
   * Mark everything transitively reachable from the queue, adding to the queue whenever a reachable
   * object brings along more reachable objects.
   */
  private fun markReachable() {
    while (true) {
      val root: Any = queue.poll() ?: break
      val reachableMembers: List<Any?> = reachableObjects(root)

      for (reachable in reachableMembers) {
        when (reachable) {
          is ProtoType -> {
            if (root is ProtoMember) {
              if (marks.mark(reachable, root)) {
                queue.add(reachable)
              }
            } else {
              if (marks.mark(reachable)) {
                queue.add(reachable)
              }
            }
          }

          is ProtoMember -> {
            if (isRetainedVersion(reachable) && marks.mark(reachable)) {
              queue.add(reachable)
            }
          }

          null -> {
            // Skip nulls.
            // TODO(jwilson): create a dedicated UNLINKED type as a placeholder.
          }

          else -> {
            throw IllegalStateException("unexpected object: $reachable")
          }
        }
      }
    }
  }

  /**
   * Returns everything reachable from `root` when traversing the graph. The returned list
   * contains instances of type [ProtoMember] and [ProtoType].
   *
   * @param root either a [ProtoMember] or [ProtoType].
   */
  private fun reachableObjects(root: Any): List<Any?> {
    val result = mutableListOf<Any?>()
    val options: Options
    var fileOptions: Options? = null

    when (root) {
      is ProtoMember -> {
        val member = root.member
        val type = schema.getType(root.type)
        val service = schema.getService(root.type)

        if (type is MessageType) {
          val field = type.field(member) ?: type.extensionField(member)
          if (field != null) {
            result.add(field.type)
            options = field.options
          } else {
            val oneOf = checkNotNull(type.oneOf(member)) { "unexpected member: $member" }
            options = oneOf.options
          }
        } else if (type is EnumType) {
          val constant = type.constant(member)
            ?: throw IllegalStateException("unexpected member: $member")
          options = constant.options
        } else if (service != null) {
          val rpc = service.rpc(member) ?: throw IllegalStateException("unexpected rpc: $member")
          result.add(rpc.requestType)
          result.add(rpc.responseType)
          options = rpc.options
        } else {
          throw IllegalStateException("unexpected member: $member")
        }
      }

      is ProtoType -> {
        if (root.isMap) {
          result.add(root.keyType)
          result.add(root.valueType)
          return result
        }

        if (root.isScalar) {
          return result // Skip scalar types.
        }

        val type = schema.getType(root)
        val service = schema.getService(root)
        fileOptions = schema.protoFile(root)!!.options

        if (type is MessageType) {
          options = type.options
          for (field in type.declaredFields) {
            result.add(get(root, field.name))
          }
          for (field in type.extensionFields) {
            result.add(get(root, field.qualifiedName))
          }
          for (oneOf in type.oneOfs) {
            result.add(get(root, oneOf.name))
            for (field in oneOf.fields) {
              result.add(get(root, field.name))
            }
          }
        } else if (type is EnumType) {
          options = type.options
          for (constant in type.constants) {
            result.add(get(type.type, constant.name))
          }
        } else if (type is EnclosingType) {
          options = type.options
        } else if (service != null) {
          options = service.options
          for (rpc in service.rpcs) {
            result.add(get(service.type, rpc.name))
          }
        } else {
          throw IllegalStateException("unexpected type: $root")
        }
      }
      else -> {
        throw IllegalStateException("unexpected root: $root")
      }
    }

    addOptions(options.fields(pruningRules).values(), result)
    if (fileOptions != null) {
      addOptions(fileOptions.fields(pruningRules).values(), result)
    }

    return result
  }

  private fun addOptions(options: Collection<ProtoMember>, result: MutableList<Any?>) {
    for (member in options) {
      // If it's an extension, don't consider the entire enclosing type to be reachable.
      if (!schema.isExtensionField(member)) {
        result.add(member.type)
      }
      result.add(member)
    }
  }
}
